test/lib/xopts.test.ts

```diff
@@ -1,29 +1,171 @@
-      const provider = new providers.Web3Provider(new FakeWeb3Provider());
-      const xopts = await XOpts.load(provider, addresses);
-      expect(xopts.options).to.have.property('list');
-      // we have static typing for the following
-      expect(xopts.options).to.not.have.property('write');
-    });
-
-    it('should work with signer', async () => {
-      const signer = new VoidSigner('0x');
-      const xopts = await XOpts.load(signer, addresses);
-      expect(xopts.options).to.have.property('list');
-      expect(xopts.options).to.have.property('write');
+  describe('unit', () => {
+    const addresses = Deployments.ganache;
+    let web3Provider: FakeWeb3Provider;
+    let provider: providers.Provider;
+
+    beforeEach(() => {
+      web3Provider = new FakeWeb3Provider();
+      provider = new providers.Web3Provider(web3Provider);
+
+      web3Provider.register('net_version', '1234', {n: 10});
+      web3Provider.register('eth_chainId', '1234', {n: 10});
+    });
+
+    describe('constructor', () => {
+      it('should work with provider', async () => {
+        const xopts = await XOpts.load(provider, addresses);
+        expect(xopts.options).to.have.property('list');
+        // we have static typing for the following
+        expect(xopts.options).to.not.have.property('write');
+      });
+
+      it('should work with signer', async () => {
+        const signer = new VoidSigner('0x');
+        const xopts = await XOpts.load(signer, addresses);
+        expect(xopts.options).to.have.property('list');
+        expect(xopts.options).to.have.property('write');
+      });
+    });
+
+    describe('globals', () => {
+      let xopts: XOpts<Provider>;
+
+      beforeEach(async () => {
+        xopts = await XOpts.load(provider, addresses);
+      });
+
+      describe('totalLiquidity', () => {
+        it('should return total USDT amount', async () => {
+          web3Provider.register('eth_call', addresses.collateral, {
+            to: addresses.optionFactory
+          });
+          web3Provider.register('eth_call', '0xfa', {to: addresses.collateral});
+          const amount = await xopts.totalLiquidity();
+          expect(amount.toString()).to.eq('250');
+        });
+      });
+
+      describe('optionMarketsCount', () => {
+        it('should return the number of options', async () => {
+          const returnValue = encode('address[]', [ZERO_ADDRESS, DAI_ADDRESS]);
+          web3Provider.register('eth_call', returnValue, {
+            to: addresses.optionFactory
+          });
+          const count = await xopts.optionMarketsCount();
+          expect(count).to.eq(2);
+        });
+      });
+    });
+  });
+
+  // NOTE: tests here use JSON RPC through HTTP
+  // the data is assumed to be the one populated by the testdata.js script
+  describe('integration', () => {
+    const web3URI = process.env.NODE_URL || 'http://localhost:8545';
+    const btcAddress = decodeBtcAddress(
+      '1BuzwRy3TMYnhQqaZMRWjkqqUbfc2wQARr',
```

Comment:
We could move that to `constants.ts`?

Reply:
We could also generate a random address with `bitcoinjs-lib`.

Reply:
Random address sounds good. I added my address here just in case someone wanted to send me money... :joy: